<details><summary>Failed to submit graph for 566e75db7f6bdd923f7da07f6cf4cdb7ba7aac40
</summary>Message:
Failed to run command: 'pbp generate from-feedstock --output /tmp/tmpglxghjma/llvmdev-feedstock.yaml --custom-metadata-file /tmp/tmpglxghjma/custom_metadata.yaml --release-name llvmdev-feedstock-36-566e75db7f6bdd923f7da07f6cf4cdb7ba7aac40 llvmdev-feedstock/recipe', with error: 'INFO:root:Found feedstock llvmdev-feedstock at /tmp/tmpglxghjma/llvmdev-feedstock/recipe
INFO:root:Build/Test graph from llvmdev-feedstock
Created by sbx_user1051 on 2026-03-31 17:19:33
Command line: /opt/conda/envs/webhook-listener/bin/pbp generate from-feedstock --output /tmp/tmpglxghjma/llvmdev-feedstock.yaml --custom-metadata-file /tmp/tmpglxghjma/custom_metadata.yaml --release-name llvmdev-feedstock-36-566e75db7f6bdd923f7da07f6cf4cdb7ba7aac40 llvmdev-feedstock/recipe

INFO:root:Found abs.yaml in /tmp/tmpglxghjma/llvmdev-feedstock
INFO:root:Creating graph from feedstock at: llvmdev-feedstock
INFO:pbpcli.utils.common:Setting env var: CONDA_SUBDIR = linux-64.
INFO:pbpcli.utils.expand_variants:Elapsed time for get_package_variants: 0.11 seconds
INFO:pbpcli.utils.expand_variants:Matrix vars: 3 unchanging vars:(490), changing vars (3): ['blas_impl', 'numpy', 'python']
INFO:pbpcli.utils.expand_variants:Elapsed time for preparation for render: 0.00 seconds
INFO:pbpcli.utils.common:Setting env var: CONDA_PKGS_DIRS = /tmp/conda_pkgs_1cwt17dx.
WARNING: failed to get package records, retrying.  exception was: Unsatisfiable dependencies for platform linux-64: {MatchSpec("libcxx-devel=14.3.0")}
Encountered problems while solving:
  - unsupported request

Could not solve for environment specs
The following package could not be installed
└─ libcxx-devel =14.3.0 * does not exist (perhaps a typo or a missing channel).
ERROR: Failed to get package records, max retries exceeded.
WARNING: Returning non-final recipe for libllvm22-22.1.2-0; one or more dependencies was unsatisfiable:
WARNING: Build: libcxx-devel
WARNING: Host: libcxx-devel
WARNING: Returning non-final recipe for llvm-tools-22-22.1.2-0; one or more dependencies was unsatisfiable:
WARNING: Returning non-final recipe for llvm-tools-22.1.2-0; one or more dependencies was unsatisfiable:
WARNING: Returning non-final recipe for llvmdev-22.1.2-0; one or more dependencies was unsatisfiable:
INFO:pbpcli.utils.common:Unsetting env var: CONDA_PKGS_DIRS
INFO:pbpcli.utils.expand_variants:Elapsed time for api.render: 97.01 seconds
INFO:pbpcli.utils.common:Unsetting env var: CONDA_SUBDIR
INFO:pbpcli.utils.common:Setting env var: CONDA_SUBDIR = linux-aarch64.
INFO:pbpcli.utils.expand_variants:Elapsed time for get_package_variants: 0.15 seconds
INFO:pbpcli.utils.expand_variants:Matrix vars: 2 unchanging vars:(490), changing vars (2): ['numpy', 'python']
INFO:pbpcli.utils.expand_variants:Elapsed time for preparation for render: 0.00 seconds
INFO:pbpcli.utils.common:Setting env var: CONDA_OVERRIDE_GLIBC = 2.28.
INFO:pbpcli.utils.common:Setting env var: CONDA_PKGS_DIRS = /tmp/conda_pkgs_s5wjrhu0.
INFO:pbpcli.utils.common:Unsetting env var: CONDA_OVERRIDE_GLIBC
INFO:pbpcli.utils.common:Unsetting env var: CONDA_PKGS_DIRS
INFO:pbpcli.utils.expand_variants:Elapsed time for api.render: 66.87 seconds
INFO:pbpcli.utils.common:Unsetting env var: CONDA_SUBDIR
INFO:pbpcli.utils.common:Setting env var: CONDA_SUBDIR = osx-arm64.
INFO:pbpcli.utils.expand_variants:Elapsed time for get_package_variants: 0.16 seconds
INFO:pbpcli.utils.expand_variants:Matrix vars: 2 unchanging vars:(493), changing vars (2): ['numpy', 'python']
INFO:pbpcli.utils.expand_variants:Elapsed time for preparation for render: 0.00 seconds
INFO:pbpcli.utils.common:Setting env var: CONDA_OVERRIDE_OSX = 12.1.
INFO:pbpcli.utils.common:Setting env var: CONDA_PKGS_DIRS = /tmp/conda_pkgs_zoubwo6q.
WARNING: Setting build platform. This is only useful when pretending to be on another platform, such as for rendering necessary dependencies on a non-native platform. I trust that you know what you're doing.
INFO:pbpcli.utils.common:Unsetting env var: CONDA_OVERRIDE_OSX
INFO:pbpcli.utils.common:Unsetting env var: CONDA_PKGS_DIRS
INFO:pbpcli.utils.expand_variants:Elapsed time for api.render: 54.50 seconds
INFO:pbpcli.utils.common:Unsetting env var: CONDA_SUBDIR
INFO:pbpcli.utils.common:Setting env var: CONDA_SUBDIR = win-64.
INFO:pbpcli.utils.expand_variants:Elapsed time for get_package_variants: 0.11 seconds
INFO:pbpcli.utils.expand_variants:Matrix vars: 2 unchanging vars:(495), changing vars (2): ['numpy', 'python']
INFO:pbpcli.utils.expand_variants:Elapsed time for preparation for render: 0.00 seconds
INFO:pbpcli.utils.common:Setting env var: CONDA_PKGS_DIRS = /tmp/conda_pkgs_eodp1t2g.
INFO:pbpcli.utils.common:Unsetting env var: CONDA_PKGS_DIRS
INFO:pbpcli.utils.common:Unsetting env var: CONDA_SUBDIR
Traceback (most recent call last):
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/conda_build/metadata.py", line 2097, in _get_contents
    rendered = template.render(environment=env)
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/jinja2/environment.py", line 1295, in render
    self.environment.handle_exception()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/jinja2/environment.py", line 942, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 25, in top-level template code
jinja2.exceptions.UndefinedError: 'cxx_compiler_version' is undefined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/envs/webhook-listener/bin/pbp", line 10, in <module>
    sys.exit(cli())
             ~~~^^
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/click/core.py", line 1442, in __call__
    return self.main(*args, **kwargs)
           ~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/click/core.py", line 1363, in main
    rv = self.invoke(ctx)
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/click/core.py", line 1830, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/click/core.py", line 1830, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/click/core.py", line 1226, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/click/core.py", line 794, in invoke
    return callback(*args, **kwargs)
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/pbpcli/cli/generate.py", line 416, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/pbpcli/cli/generate.py", line 643, in from_feedstock
    build_configs, test_configs = expand_build_matrix(
                                  ~~~~~~~~~~~~~~~~~~~^
        feedstock.recipe_dir_path, src_channels, parsed_user_variants
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/pbpcli/cli/generate.py", line 82, in expand_build_matrix
    _targets, build_configs, test_configs = variants_for_feedstock_and_platform(
                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        recipe_dir_path, platform, src_channels, user_variants
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/pbpcli/utils/expand_variants.py", line 566, in variants_for_feedstock_and_platform
    metadata_tuples = api.render(
        parsed.recipe,
    ...<2 lines>...
        variants=user_variants,
    )
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/conda_build/api.py", line 68, in render
    return render_metadata_tuples(
        metadata_tuples,
    ...<3 lines>...
        bypass_env_check=bypass_env_check,
    )
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/conda_build/render.py", line 1097, in render_metadata_tuples
    for od, om in meta.get_output_metadata_set(
                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        permit_unsatisfiable_variants=permit_unsatisfiable_variants,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        permit_undefined_jinja=not finalize,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        bypass_env_check=bypass_env_check,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ):
    ^
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/conda_build/metadata.py", line 2781, in get_output_metadata_set
    conda_packages = finalize_outputs_pass(
        ref_metadata,
    ...<3 lines>...
        bypass_env_check=bypass_env_check,
    )
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/conda_build/metadata.py", line 1063, in finalize_outputs_pass
    output_d = om.get_rendered_output(metadata.name()) or {
               ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/conda_build/metadata.py", line 2867, in get_rendered_output
    for output_ in self.get_rendered_outputs_section(
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        permit_undefined_jinja=permit_undefined_jinja, variant=variant
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ):
    ^
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/conda_build/metadata.py", line 2849, in get_rendered_outputs_section
    self._get_contents(
    ~~~~~~~~~~~~~~~~~~^
        permit_undefined_jinja=permit_undefined_jinja,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<3 lines>...
        variant=variant,
        ^^^^^^^^^^^^^^^^
    )
    ^
  File "/opt/conda/envs/webhook-listener/lib/python3.13/site-packages/conda_build/metadata.py", line 2107, in _get_contents
    raise CondaBuildUserError(
        f"Failed to render jinja template in {self.meta_path}:\n{str(ex)}"
    )
conda_build.exceptions.CondaBuildUserError: Failed to render jinja template in /tmp/tmpglxghjma/llvmdev-feedstock/recipe/meta.yaml:
'cxx_compiler_version' is undefined
'
aws lambda request id: b67b8a7b-f2ba-58fd-8e78-e83df6eaf466
[View in Grafana](https://grafana.anacondaconnect.com/explore?schemaVersion=1&panes=%7B%2286g%22:%7B%22datasource%22:%22loki-central%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7B__aws_cloudwatch_log_group%3D%5C%22%2Faws%2Flambda%2Fpbp-graph-submitter-prod%5C%22%7D%20%7C%20json%20%7C%20requestId%20%3D%20%60b67b8a7b-f2ba-58fd-8e78-e83df6eaf466%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22loki-central%22%7D,%22editorMode%22:%22builder%22,%22direction%22:%22backward%22%7D%5D,%22range%22:%7B%22from%22:%22now-3h%22,%22to%22:%22now%22%7D,%22panelsState%22:%7B%22logs%22:%7B%22columns%22:%7B%220%22:%22Time%22,%221%22:%22Line%22%7D,%22visualisationType%22:%22table%22,%22labelFieldName%22:%22labels%22%7D%7D%7D%7D&orgId=1)</details>

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -2,6 +2,8 @@
 {% set major_ver = version.split(".")[0] %}
 {% set tail_ver = version.split(".")[-1] %}
 {% set maj_min = major_ver ~ "." ~ version.split(".")[1] %}
+# Stage 0: pin libcxx-devel to old version (Jinja renders before selectors, so need default for non-osx)
+{% set libcxx_devel_ver = cxx_compiler_version | default(version) %}
 
 # as of LLVM 19, we expect an "-rcX" suffix for the release candidates
 {% set extra = "-" ~ tail_ver if tail_ver not in "0123456789" else "" %}
@@ -31,7 +33,7 @@ requirements:
     - cmake
     - ninja
     - python
-    - libcxx-devel {{ cxx_compiler_version }}    # [osx]  # Stage 0: use old libcxx (change to {{ version }} in Stage 2)
+    - libcxx-devel {{ libcxx_devel_ver }}    # [osx]  # Stage 0: uses cxx_compiler_version (change to {{ version }} in Stage 2)
     - m2-sed        # [win]
     - m2-grep       # [win]
     - m2-coreutils  # [win]  # for cp, echo, head, tail, uniq, etc.
@@ -43,7 +45,7 @@ requirements:
     - libxml2 {{ libxml2 }}
     - zlib {{ zlib }}
     - zstd {{ zstd }}
-    - libcxx-devel {{ cxx_compiler_version }}  # [osx]  # Stage 0: use old libcxx (change to {{ version }} in Stage 2)
+    - libcxx-devel {{ libcxx_devel_ver }}  # [osx]  # Stage 0: uses cxx_compiler_version (change to {{ version }} in Stage 2)
 
 outputs:
   # Contains everything
@@ -62,9 +64,9 @@ outputs:
         - ninja
         - python
         - m2-sed                  # [win]
-        - libcxx-devel {{ cxx_compiler_version }}  # [osx]  # Stage 0: use old libcxx (change to {{ version }} in Stage 2)
+        - libcxx-devel {{ libcxx_devel_ver }}  # [osx]  # Stage 0: uses cxx_compiler_version (change to {{ version }} in Stage 2)
       host:
-        - libcxx-devel {{ cxx_compiler_version }}  # [osx]  # Stage 0: use old libcxx (change to {{ version }} in Stage 2)
+        - libcxx-devel {{ libcxx_devel_ver }}  # [osx]  # Stage 0: uses cxx_compiler_version (change to {{ version }} in Stage 2)
         - {{ pin_subpackage("libllvm" ~ major_ver, exact=True) }}
         - {{ pin_subpackage("libllvm-c" ~ major_ver, exact=True) }}     # [win]
         - {{ pin_subpackage("llvm-tools", exact=True) }}
@@ -75,7 +77,7 @@ outputs:
         - {{ pin_subpackage("libllvm-c" ~ major_ver, exact=True) }}     # [win]
         - {{ pin_subpackage("llvm-tools", exact=True) }}
         # we need to do this manually because clang_bootstrap has no run-export
-        - libcxx >={{ cxx_compiler_version }}                           # [osx]
+        - libcxx >={{ libcxx_devel_ver }}                           # [osx]
       # Stage 0: run_constrained disabled (re-enable in Stage 2)
       # run_constrained:
       #   - llvm        {{ version }}
@@ -115,15 +117,15 @@ outputs:
         - ninja                    # [not win]
         - cmake                    # [not win]
         - python                   # [not win]
-        - libcxx-devel {{ cxx_compiler_version }}  # [osx]  # Stage 0: use old libcxx (change to {{ version }} in Stage 2)
+        - libcxx-devel {{ libcxx_devel_ver }}  # [osx]  # Stage 0: uses cxx_compiler_version (change to {{ version }} in Stage 2)
       host:
-        - libcxx-devel {{ cxx_compiler_version }}  # [osx]  # Stage 0: use old libcxx (change to {{ version }} in Stage 2)
+        - libcxx-devel {{ libcxx_devel_ver }}  # [osx]  # Stage 0: uses cxx_compiler_version (change to {{ version }} in Stage 2)
         - libxml2 {{ libxml2 }} # [unix]
         - zlib {{ zlib }}
         - zstd {{ zstd }}
       run:
         # we need to do this manually because clang_bootstrap has no run-export
-        - libcxx >={{ cxx_compiler_version }}  # [osx]
+        - libcxx >={{ libcxx_devel_ver }}  # [osx]
     test:
       commands:
         # old style
@@ -175,16 +177,16 @@ outputs:
         - cmake
         - ninja
         - python >=3
-        - libcxx-devel {{ cxx_compiler_version }}  # [osx]  # Stage 0: use old libcxx (change to {{ version }} in Stage 2)
+        - libcxx-devel {{ libcxx_devel_ver }}  # [osx]  # Stage 0: uses cxx_compiler_version (change to {{ version }} in Stage 2)
       host:
-        - libcxx-devel {{ cxx_compiler_version }}  # [osx]  # Stage 0: use old libcxx (change to {{ version }} in Stage 2)
+        - libcxx-devel {{ libcxx_devel_ver }}  # [osx]  # Stage 0: uses cxx_compiler_version (change to {{ version }} in Stage 2)
         - {{ pin_subpackage("libllvm" ~ major_ver, exact=True) }}
         - zlib
         - zstd
       run:
         - {{ pin_subpackage("libllvm" ~ major_ver, exact=True) }}
         # we need to do this manually because clang_bootstrap has no run-export
-        - libcxx >={{ cxx_compiler_version }}  # [osx]
+        - libcxx >={{ libcxx_devel_ver }}  # [osx]
     test:
       commands:
         - $PREFIX/bin/llc-{{ major_ver }} -version                               # [not win]
@@ -208,7 +210,7 @@ outputs:
         - cmake
         - ninja
         - python
-        - libcxx-devel {{ cxx_compiler_version }}  # [osx]  # Stage 0: use old libcxx (change to {{ version }} in Stage 2)
+        - libcxx-devel {{ libcxx_devel_ver }}  # [osx]  # Stage 0: uses cxx_compiler_version (change to {{ version }} in Stage 2)
       host:
         - {{ pin_subpackage("libllvm" ~ major_ver, exact=True) }}
         - {{ pin_subpackage("llvm-tools-" ~ major_ver, exact=True) }}   # [not win]
@@ -253,9 +255,9 @@ outputs:
         - {{ stdlib('c') }}
         - cmake
         - ninja
-        - libcxx-devel {{ cxx_compiler_version }}  # [osx]  # Stage 0: use old libcxx (change to {{ version }} in Stage 2)
+        - libcxx-devel {{ libcxx_devel_ver }}  # [osx]  # Stage 0: uses cxx_compiler_version (change to {{ version }} in Stage 2)
       host:
-        - libcxx-devel {{ cxx_compiler_version }}  # [osx]  # Stage 0: use old libcxx (change to {{ version }} in Stage 2)
+        - libcxx-devel {{ libcxx_devel_ver }}  # [osx]  # Stage 0: uses cxx_compiler_version (change to {{ version }} in Stage 2)
         - {{ pin_subpackage("libllvm" ~ major_ver, exact=True) }}
         - zlib {{ zlib }}
         - zstd {{ zstd }}

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -2,8 +2,10 @@
 {% set major_ver = version.split(".")[0] %}
 {% set tail_ver = version.split(".")[-1] %}
 {% set maj_min = major_ver ~ "." ~ version.split(".")[1] %}
-# Stage 0: pin libcxx-devel to old version (Jinja renders before selectors, so need default for non-osx)
-{% set libcxx_devel_ver = cxx_compiler_version | default(version) %}
+# Stage 0: pin libcxx-devel to old LLVM version for bootstrap
+# Jinja renders before selectors, so this must be valid on all platforms
+# Change to `version` in Stage 2 for self-hosted build
+{% set libcxx_devel_ver = "21.1.8" %}
 
 # as of LLVM 19, we expect an "-rcX" suffix for the release candidates
 {% set extra = "-" ~ tail_ver if tail_ver not in "0123456789" else "" %}

